tss-esapi/src/context/tpm_commands/attestation_commands.rs

```diff
@@ -12,7 +12,45 @@ use std::convert::TryFrom;
 use std::ptr::null_mut;
 
 impl Context {
-    // Missing function: Certify
+    /// Prove that an object is loaded in the TPM
+    ///
+    /// # Errors
+    /// * if the qualifying data provided is too long, a `WrongParamSize` wrapper error will be returned
+    pub fn certify(
+        &mut self,
+        object_handle: ObjectHandle,
+        signing_key_handle: KeyHandle,
+        qualifying_data: &Data,
+        signing_scheme: TPMT_SIG_SCHEME,
+    ) -> Result<(TPM2B_ATTEST, Signature)> {
+        let mut certify_info = null_mut();
+        let mut signature = null_mut();
+        let ret = unsafe {
+            Esys_Certify(
+                self.mut_context(),
+                object_handle.into(),
+                signing_key_handle.into(),
+                self.required_session_1()?,
+                self.optional_session_2(),
```

Comment:
I think this one is required as well, as both `object_handle` and `signing_key_handle` need separate authorizations.

Reply:
It's optional because signing_key_handle may be Null. From the "Trusted Platform Module Library, Part 3: Commands" doc:

`
Type                | Name           | Description
TPMI_ST_COMMAND_TAG | tag            | TPM_ST_SESSIONS
UINT32              | commandSize    |
TPM_CC              | commandCode    | TPM_CC_Certify
TPMI_DH_OBJECT      | @objectHandle  | handle of the object to be certified / Auth Index: 1 / Auth Role: ADMIN
TPMI_DH_OBJECT+     | @signHandle    | handle of the key used to sign the attestation structure / Auth Index: 2 / Auth Role: USER
TPM2B_DATA          | qualifyingData | user provided qualifying data
TPMT_SIG_SCHEME+    | inScheme       | signing scheme to use if the scheme for signHandle is TPM_ALG_NULL
`

Where + is defined to mean:
> A Type decoration – When appended to a value in the Type column of a command, this symbol
indicates that the parameter is allowed to use the “null” value of the data type (see "Conditional
Types" in TPM 2.0 Part 2). The null value is usually TPM_RH_NULL for a handle or
TPM_ALG_NULL for an algorithm selector

Reply:
Ah, yes! Also didn't notice that

> If signHandle  is  TPM_RH_NULL,  the  TPMS_ATTEST  structure  is  returned  and  signature  is  a  NULL Signature.

That actually raises a few other questions.

First would be, is it possible at the moment to create a null `KeyHandle`? I can't see a constructor for it, so it's only through a conversion from `ESYS_TR`. Would it make sense to add a static value for null, kinda like [this](https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi/src/structures/buffers/public/rsa.rs#L231)?

Second, should we check here whether `signing_key_handle` is null or not before calling either `optional_session_2` or `required_session_2`?

Reply:
I ended up making the both required as originally suggested as although the second handle is optional, upon testing with a software TPM and re-reading the spec, it seems the session is still required even if the handle is NULL.